### PR TITLE
Try to fix test_storage_s3: crash in WriteBufferFromS3

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`std::terminate` was called if there is an error writing data into s3.

### CODE REVIEW

File: src/IO/WriteBufferFromS3.cpp [hunk 1]
```diff
@@ -107,7 +107,14 @@ void WriteBufferFromS3::finalizeImpl()
...
 WriteBufferFromS3::~WriteBufferFromS3()
 {
-    finalizeImpl();
+    try
+    {
+        finalizeImpl();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
```
Comment (azat) [inline src/IO/WriteBufferFromS3.cpp:116]: But this will hide write error, does not looks right...

Comment (alexey-milovidov) [reply src/IO/WriteBufferFromS3.cpp:116]: Let's merge this PR first and another PR later.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@azat FYI another idea is to make std::terminate in debug and logging in release.

**Comment by azat** [comment]
>@azat FYI another idea is to make std::terminate in debug and logging in release.

@alexey-milovidov I have similar thoughts, but decided not to, since write error may hide data loss.

**Comment by azat** [comment]
@vdimir can you provide a stack trace for uncaught exception?

**Comment by vdimir** [comment]
> @vdimir can you provide a stack trace for uncaught exception?

@azat  https://pastebin.com/raw/1N0na6X4
(from here https://clickhouse-test-reports.s3.yandex.net/21457/e69f7c7a94d7c0a5776cc408b99a028d3eeae06b/integration_tests_flaky_check_(asan).html#fail1  see file `test_storage_s3/_instances/dummy/logs/clickhouse-server.err.log` from `test_dir.tar`)

**Comment by azat** [comment]
>(from here https://clickhouse-test-reports.s3.yandex.net/21457/e69f7c7a94d7c0a5776cc408b99a028d3eeae06b/integration_tests_flaky_check_(asan).html#fail1 see file test_storage_s3/_instances/dummy/logs/clickhouse-server.err.log from test_dir.tar)

@vdimir thanks!

I will try to remove that try/catch to see if test 100% fails, and later push some fixes - #22208

**Pull request merged**